### ISSUE TITLE
Freeze in partition_by_union with non-overlapping colors

### ISSUE THREAD

**Issue by elliottslaughter** [StanfordLegion/legion#452]
@syamajala recently showed me a case where we hang when we take the union of partitions `A | B` if `A.colors` and `B.colors` is non-overlapping (i.e. each of `A.colors` and `B.colors` contain points not in the other). It probably also occurs in other cases too, this is just the reproducer we have right now.

What's the expected behavior in this case? Clearly, it shouldn't freeze, but is this expected to work?

```
import "regent"

task main()
  var mat = region(ispace(int2d, {10, 10}), double)

  var colors = ispace(int2d, {1, 1}, {-1, -1})
  var part = partition(equal, mat, colors)
  regentlib.c.printf("vol: %d\n", part[{-1, -1}].ispace.volume)

  var colors2 = ispace(int2d, {1, 1})
  var part2 = partition(equal, mat, colors2)
  regentlib.c.printf("vol2: %d\n", part2[{0, 0}].ispace.volume)

  var up = part | part2
  regentlib.c.printf("vol3: %d\n", up[{0, 0}].ispace.volume)
end
regentlib.start(main)
```

To be clear, when Regent detects that the color spaces are not equal, it internally unions them before passing this on to the runtime: https://github.com/StanfordLegion/legion/blob/482e188d06bf9050b5d5a85fd0050c22374cf0ee/language/src/regent/codegen.t#L6952-L6957

**Comment by lightsighter** [comment]
@elliottslaughter How do you compute this parameter to create_partition_by_union:
https://gitlab.com/StanfordLegion/legion/blob/master/runtime/legion/legion_c.cc#L1187

**Comment by elliottslaughter** [comment]
It's in the highlighted code up above. I'm taking the two argument color spaces and passing them to `legion_index_space_union` to get the union color space.

**Comment by lightsighter** [comment]
What happens if you run the code with:
var up = part & part2

**Comment by syamajala** [comment]
That also freezes.

**Comment by lightsighter** [comment]
Actually I think this is a compiler bug. I think the compiler is unioning the color spaces when it should be intersecting them. @elliottslaughter will have to say for sure. But at least the runtime is expecting all these kinds dependent partition operations to happen on intersected color spaces which you can't control from user code in Regent.

**Comment by syamajala** [comment]
@magnatelee can you look at this?

**Comment by elliottslaughter** [comment]
In my opinion the Regent interpretation is defensible; why would you ask
for a union operator but then expect the color spaces to be intersected?

But either way, it should be trivial to fix in Regent. The change to get it
to use intersections will be one line. The main thing to keep in mind is
that Seshu's code will need to change, in that all the regions on either
side of the union operator will have to be created up front, and the ones
that currently don't exist in one side or the other will just be made empty.

The one thing I don't think we should do is allow the Regent and Legion
behaviors to diverge. That's just asking for confusion.

**Comment by syamajala** [comment]
I found that even if one color space contains the other you can't assign a partition to itself. I'll try to see if I can come up with a small example.

**Comment by streichler** [comment]
Assigning this to @elliottslaughter and @lightsighter to agree on the right way for both Legion and Regent to deal with this.

**Comment by elliottslaughter** [comment]
I can make Regent conform to Legion's interpretation. It wouldn't solve @syamajala's problem but I think I've convinced myself we need a different approach there anyway at this point.

**Comment by elliottslaughter** [comment]
@lightsighter Did we decide to go with Legion or Regent's interpretation on this one?

**Comment by lightsighter** [comment]
I still don't think Regent's approach is defensible. Consider the case of subtracting one partition from another: what do you do if the colors spaces of the two partitions don't match? You can't subtract a subregion from nothing. In my opinion Legion's approach is the only one that is well defined: regardless of the set operation being performed, we only perform it on the points that exist in the color space of both partition operands. It was even documented like this in legion.h before Regent implemented the operators. :)

**Comment by streichler** [comment]
I'm bringing another piece into this because I think they're coupled:  If I have a partition in which many of the subregions are empty (e.g. a 2-D tiled partition of a sparse matrix that has few off-diagonal entries), what is the space-efficient way to represent that partition in Legion?  Is it:
a) a full (i.e. dense 2-D rectangle) color space for which Legion (efficiently) knows which colors actually have (non-empty) index spaces allocated to track the subspace, or 
b) a sparse color space, allowing Legion to always pay for an IndexSpace for each point in the color space

If the answer is (a), then I'm fine with Legion's semantics.  Set operations will use two partitions with the same dense 2-D color space even though their sparsity patterns may be different, and the right thing will happen.
But if the answer is (b), then I think Regent's semantics are needed.  Otherwise, any set operation on two partitions with the same "conceptual color space" (i.e. the colors of possible subregions before empty-ness is considered) will require the application to do the corresponding set operations on the two partitions' color spaces first to get the desired behavior.

**Comment by lightsighter** [comment]
Except for (b) in the common case which is Legion will recognize that the names for the color spaces of the two partitions are the same and then we don't need to do the intersection explicitly.

**Comment by streichler** [comment]
My point was that with (b), it'd be the application's problem to use a sparse color space for each of the two partitions, and then they wouldn't be the same from Legion's perspective.  The application being able to (space-efficiently) use the same color space for both partitions, regardless of which subregions are empty in each partition, is choice (a).

**Comment by lightsighter** [comment]
Legion always requires a color space to be an index space (literally a realm index space). Whether that color space is space efficient or not depends on realm's implementation of index spaces, not anything that Legion does. 

The Legion interface for all these operations currently requires that the application always provide the color space for the new partition, so we could get rid of the "intersection" color space semantics, but if we do I want to know what the runtime should do when it iterates over the color space provided by the user and then can't find the corresponding sub-region for one of the colors in one of the input partitions.

**Comment by streichler** [comment]
The scalability concern applies at several levels.  Yes, individual Realm index spaces need storage and computation costs that are no larger than linear in the number of points in the set, but there are other places where super-linear costs can come in.  For example, the cross-product intersection of two O(N) partitions conceptually yields an O(N^2) color space but will have only O(N) non-empty intersection subregions.  If we create a Realm index space for every possible subregion, that's a scalability issue.  If we don't, then there are a few different ways to prune out the empty subregions, and they can be classified by whether they are the application's problem to pre-compute the sparse color space that covers just the non-empty intersections (which becomes case (b) above), or whether it's Legion's problem to take advantage of the sparsity in the intersection and present the original O(N^2) (but dense, so cheap) color space to the application.

It may make sense to table this discussion until @lightsighter and I can do a round in front of a white board.

**Comment by streichler** [comment]
Ok, after several more rounds of discussion, I believe the "regent interpretation" is the one we should go with.  Since the Legion operations demand that the application pre-compute the desired color space for the result of a partitioning operation (e.g. https://gitlab.com/StanfordLegion/legion/blob/master/runtime/legion.h#L3673), Legion should respect that request.  In particular, rather than ignoring requested colors that do not appear in one or the other of the two input partitions (i.e. 
> The zip operation is only applied 
> to the points contained in the intersection of the two color
> spaces.

), I think we should change the semantics to not ignore colors and assume that missing colors in either of the input partitions correspond to empty subregions.  This would apply to all Legion partitioning operations that accept a color space from the application. 
@elliottslaughter , @lightsighter , thoughts?

**Comment by lightsighter** [comment]
So the rule is that even empty sub-regions in the output partition are always represented? For example, in a partition difference, if there is no color for the lhs sub-region then that gets converted to empty and the resulting partition still has an explicitly empty sub-region for that point in the color space.

**Comment by streichler** [comment]
> So the rule is that even empty sub-regions in the output partition are always represented?
Yes, subregions (empty or not) exist for exactly the colors requested by the application.
